add pinot upsert features to pinot common

jamesyfshao opened the pull request:
This diff will create necessary pinot schema/constant value changes to support pinot upsert features.

The following will be changed:

1. add IngestionModeConfig field to schema and table definition:
This field will contains several sub-fields:
   a. ingestionMode: 
ingestion mode could be either upsert/append.
  b. primaryKeys:
a list of string that indicate which columns would be use as primary key of the current upsert table. However we are going to enforce primary key to be one only only one column. In the future we can explore the support for multiple primary keys
  c. offsetKey:
a string that indicate which column would be use as offset of the current pinot upsert table.

design doc for pinot upsert feature:

https://cwiki.apache.org/confluence/display/PINOT/%5BProposal%5D+Pinot+upsert+design+doc
this change focus on the changes in pinot common repo

kishoreg commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java:
```diff
@@ -75,6 +79,9 @@
   private List<FieldConfig> _fieldConfigList;
 
+  @JsonPropertyDescription(value = "The update semantic of the table, either append or upsert, default as append")
+  private UpdateSemantic _updateSemantic;
```
rename to IngestMode or just have a boolean enableUpsert?

Jackie-Jiang replied:
Is this for real-time? We already have push type of either APPEND or REFRESH for offline; aggregateMetrics for real-time. What does UPSERT stand for?

jamesyfshao replied:
@kishoreg I am thinking if using enum here will be more flexible in the case of future features changes or whatnot. Boolean seems to be only limiting two options and it might limit what other changes we might want to do with pinot. Please let me what you think. ingestionMode definitely sounds more explicit than what I am using here and I will definite consider about this.

@Jackie-Jiang yes. So basically it is previous old design about we want to support update in pinot realtime ingestion by adding override by primary key (eg, a second message in Kafka with the same primary key represents the update of existing key instead of two records with the same primary key). Upsert stands for this new ingestion semantic (insert if no duplicated primary key, update if duplicate primary key). This feature will be applying to realtime table only

kishoreg commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableNameBuilder.java:
```diff
@@ -132,4 +132,19 @@ public static boolean isOfflineTableResource(String resourceName) {
+   * @param type the type of the table for it to fill in if the type info is missing
+   * @return the table type name with the type info
+   */
+  public static String ensureTableNameWithType(String tableName, TableType type) {
```
where is this used?

Jackie-Jiang replied:
Please use `TableNameBuilder.forType(type).tableNameWithType(tableName)`

jamesyfshao replied:
removed this method, @kishoreg this method is used in other components that I will submit PR later

kishoreg commented on pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerGauge.java:
```diff
@@ -26,7 +26,8 @@
 public enum BrokerGauge implements AbstractMetrics.Gauge {
-  QUERY_QUOTA_CAPACITY_UTILIZATION_RATE("tables", false), NETTY_CONNECTION_CONNECT_TIME_MS("nettyConnection", true);
+  QUERY_QUOTA_CAPACITY_UTILIZATION_RATE("tables", false), NETTY_CONNECTION_CONNECT_TIME_MS("nettyConnection", true),
+  TABLE_MIN_LOW_WATER_MARK("tables", false);
```
what does this represent?

jamesyfshao replied:
It represents how many table pinot broker is collecting the metrics for. The exact usage of the metrics will show up in later diffs

kishoreg commented on pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:
```diff
@@ -59,13 +65,28 @@
+
+  // upsert related config
+  // primary key refers to the column name that is the primary key of this upsert table
+  private String _primaryKey;
```
what if the primary key is made up of two columns?

kishoreg replied:
or multiple

jamesyfshao replied:
right now we don't have support for multiple primary keys. Do you think it will be an important feature to add multiple primary key support right now or you think it is fine to postpone it to later diff?

kishoreg replied:
Better to design it for multiple, even if the implementation can handle only one primary key

ChethanUK replied:
> don't have support for multiple primary keys

There will be some cases where multiple keys make sense right? Like in clickstream: `userId and timestamp`

jamesyfshao replied:
@kishoreg @ChethanUK I have changed the primary key in schema config to be a list instead of a single string. For now we are still performing check to ensure that primary key will be one and only one. However we should be able to change that once we have mulitple-primary-key support in place

kishoreg commented on pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:
```diff
@@ -59,13 +65,28 @@
+  // primary key refers to the column name that is the primary key of this upsert table
+  private String _primaryKey;
+  // offset key refers to the column name that we are going to store the offset value to
+  private String _offsetKey;
```
do we need all these things in Schema? What if we create a UpsertConfig in TableConfig to provide all the necessary information.

jamesyfshao replied:
definitely would be better this way, will submit a change later to update this

kishoreg commented on pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:
```diff
@@ -631,4 +751,50 @@ public int hashCode() {
+
+
+  @JsonIgnore
+  public static byte[] getByteArrayFromField(Object value, DimensionFieldSpec fieldSpec) {
```
why is this needed? it will be great if we can avoid static methods in config objects. We can probably move this SchemaUtil

jamesyfshao replied:
done.

It is used for converting a pinot column value into the appropriate binary representation for upsert primary key field. (make sure all primary key will become a byte array)

kishoreg commented on pom.xml:
```diff
@@ -132,6 +132,9 @@
     <log4j.version>2.11.2</log4j.version>
     <netty.version>4.1.42.Final</netty.version>
 
+    <!-- dependency for upsert component -->
```
Do we need this at compile time?

jamesyfshao replied:
moved to submodules

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java:
```diff
@@ -30,17 +30,21 @@
 import java.util.List;
 import java.util.Map;
 import javax.annotation.Nullable;
+import javax.validation.constraints.Null;
```
Wrong import

jamesyfshao replied:
issue with merging, fixed. Thanks for pointing it out

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java:
```diff
@@ -354,6 +387,7 @@ public ZNRecord toZNRecord()
     }
 
     ZNRecord znRecord = new ZNRecord(_tableName);
+    simpleFields.put(UPDATE_SEMANTIC_CONFIG_KEY, _updateSemantic.toString());
```
This can cause NPE

jamesyfshao replied:
updated to follow the convention of rest of the code

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java:
```diff
@@ -21,16 +21,17 @@
 import com.google.common.annotations.VisibleForTesting;
 import com.yammer.metrics.core.MetricName;
 import com.yammer.metrics.core.MetricsRegistry;
+import org.apache.pinot.common.Utils;
```
Wrong order, please revert

jamesyfshao replied:
| Is this for real-time? We already have push type of either APPEND or REFRESH for offline; aggregateMetrics for real-time. What does UPSERT stand for?

replied in earlier comment, basically design for supporting update/insert in pinot realtime ingestion

jamesyfshao commented:
> Can you please link the design doc in the pr description so that it is easier to understand the concept of upsert?
> Since this pr is touching the core configs, we need to spend some time discussing how to design the config keys because they are very hard to change in the future

Linked the design doc in the description. also [google doc design doc](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=4ypfJoFtqjSdua_7dziWCQ&ved=0ahUKEwjVhpqIybToAhXDHWIKHWFcBVs4ABABKAAwAA&url=https://drive.google.com/a/uber.com/open?id%3D1SFFir7ByxCff-aVYxQeTHpNhPXeP5q7P4g_6O2iNGgU%26usp%3Dchrome_omnibox&usg=AOvVaw3KEdQYFMPa7tgpOn01R5ol) with comments. We have discussed this a long time back but I definitely open to go through it again as things have already changed. Do you think we should do some remote meeting sync or something else?

kishoreg commented:
One concern here would be that the update related configuration appears in multiple places. One thing that can simplify this PR is to add a UpdateConfig under TableConfig and move everything under that. That allows you to mark that config as beta/unstable and iterate quickly
- updatesemantic (or the new name)
- primarykey
- offsetkey
-

ChethanUK commented:
> also [google doc design doc](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=4ypfJoFtqjSdua_7dziWCQ&ved=0ahUKEwjVhpqIybToAhXDHWIKHWFcBVs4ABABKAAwAA&url=https://drive.google.com/a/uber.com/open?id%3D1SFFir7ByxCff-aVYxQeTHpNhPXeP5q7P4g_6O2iNGgU%26usp%3Dchrome_omnibox&usg=AOvVaw3KEdQYFMPa7tgpOn01R5ol) with comments.

It's asking uber login. Can you change share settings?

jamesyfshao commented:
> > also [google doc design doc](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=4ypfJoFtqjSdua_7dziWCQ&ved=0ahUKEwjVhpqIybToAhXDHWIKHWFcBVs4ABABKAAwAA&url=https://drive.google.com/a/uber.com/open?id%3D1SFFir7ByxCff-aVYxQeTHpNhPXeP5q7P4g_6O2iNGgU%26usp%3Dchrome_omnibox&usg=AOvVaw3KEdQYFMPa7tgpOn01R5ol) with comments.
> 
> It's asking uber login. Can you change share settings?
> 
> > remote meeting sync or something else
> 
> @kishoreg At least monthly meeting and sync up on project status will be helpful to all...

Yeah our google doc account sharing setting is kind of wanky that it doesn't allow blanket share with everyone outside our organization. I hate it but it is really out of my control. However, just submit a request to view/comment and I can add you as soon as I see it.

We originally plan to have a meeting earlier this month but the pandemic kind of mess up with all of our plans. Maybe we can have a work-group specific remote meeting (eg, upsert meeting with kishore/jackie/chethan etc) in the meantime and do the pinot committee meeting later 

@Jackie-Jiang @kishoreg @ChethanUK let me know if you feel meeting is necessary or you prefer me to update the schema first and see if a remote meeting is necessary

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:
```diff
@@ -308,6 +320,23 @@ public ServerType getServerType() {
     public static final String PREFIX_OF_CONFIG_OF_PINOT_CRYPTER = "crypter";
   }
 
+  public static class Grigio {
```
Can we change the name of this class like I had suggested before?

mcvsubbu commented on pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:
```diff
@@ -378,4 +407,23 @@ public ServerType getServerType() {
   }
+
+  public enum UpdateSemantic {
+    APPEND,
```
Please add comments on each of these. What does APPEND mean, vs what UPSERT means. Feel free to point to the design doc

kishoreg commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java:
```diff
@@ -470,6 +505,18 @@ public void setInstanceAssignmentConfigMap(
+    _updateSemantic = updateSemantic;
+  }
+
+  public boolean isTableForUpsert() {
```
let's avoid adding additional utility methods to config classes if possible

jamesyfshao commented:
@kishoreg @mcvsubbu I have updated the diff per suggestion you guys gave in the last meeting, please feel free to take a look

@Jackie-Jiang I have updated the diff to make sure #1: it is only related table config changes required by pinot upsert feature (see [doc](https://docs.google.com/document/d/1SFFir7ByxCff-aVYxQeTHpNhPXeP5q7P4g_6O2iNGgU/edit?usp=sharing) for details) and #2 push other pinot-common related changes for pinot upsert features to later PRs

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:
```diff
@@ -0,0 +1,152 @@
+  // validFrom key refers to the column name that we are going to store the validFrom value to
+  private String _validUntilKey;
+
+  public static final String UPSERT_TABLE_CONFIG_VALUE = "upsert";
```
we should not need these two for now

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:
```diff
@@ -0,0 +1,152 @@
+
+    if (UPSERT_TABLE_CONFIG_VALUE.equals(_ingestionMode)) {
+      Preconditions.checkState(_primaryKeys.size() == 1,
+          "pinot upsert require one and only one primary key");
```
```suggestion
          "Upsert feature supports only one primary key");
```

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:
```diff
@@ -0,0 +1,152 @@
+      Preconditions.checkState(_primaryKeys.size() == 1,
+          "pinot upsert require one and only one primary key");
+      Preconditions.checkState(StringUtils.isNotEmpty(_offsetKey),
+          "pinot upsert require one offset key");
```
```suggestion
          "Upsert feature requires \"offsetColumn\" to be set");
```

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:
```diff
@@ -0,0 +1,152 @@
+      Preconditions.checkState(StringUtils.isNotEmpty(_offsetKey),
+          "pinot upsert require one offset key");
+      Preconditions.checkState(StringUtils.isNotEmpty(_validFromKey),
+          "pinot upsert require one validFrom key");
```
```suggestion
          "Upsert feature requires \"validFromColumn\" to be set");
```

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:
```diff
@@ -0,0 +1,152 @@
+      Preconditions.checkState(StringUtils.isNotEmpty(_validFromKey),
+          "pinot upsert require one validFrom key");
+      Preconditions.checkState(StringUtils.isNotEmpty(_validUntilKey),
+          "pinot upsert require one validUntil key");
```
```suggestion
          "Upsert feature requires \"validUntilColumn\"  to be set");
```

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/TableConfig.java:
```diff
@@ -44,6 +45,7 @@
   public static final String QUERY_CONFIG_KEY = "query";
   public static final String INSTANCE_ASSIGNMENT_CONFIG_MAP_KEY = "instanceAssignmentConfigMap";
   public static final String FIELD_CONFIG_LIST_KEY = "fieldConfigList";
+  public static final String INGESTION_MODE_CONFIG_KEY = "ingestionModeConfig";
```
```suggestion
  public static final String UPSERT_CONFIG_KEY = "upsertConfig";
```

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:
```diff
@@ -0,0 +1,152 @@
+
+@SuppressWarnings("unused")
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class IngestionModeConfig {
```
+1. I think this config is only for upsert feature, and not applicable for general ingestion. When this config exist, we can assume the upsert feature is on.

Jackie-Jiang replied:
Also, please extends BaseJsonConfig, which handles the Json ser-de and common methods like hashCode(), equals() and toString(). You can refer to `RoutingConfig` which is similar to this one.

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:
```diff
@@ -0,0 +1,152 @@
+  // for other value it would be append value for now
+  private String _ingestionMode;
+  // primary key refers to the column name that is the primary key of this upsert table
+  private List<String> _primaryKeys;
```
+1. You can also make it final. Same for other places

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:
```diff
@@ -0,0 +1,152 @@
+  @JsonCreator
+  public IngestionModeConfig(
+      @JsonProperty(value="ingestionMode") @Nullable String ingestionMode,
+      @JsonProperty(value="primaryKeys") @Nullable List<String> primaryKeys,
```
If this config is for upsert only, I think all arguments are required

jamesyfshao commented:
@mcvsubbu @Jackie-Jiang updated per your feedback and make the config to be used only if the table is for upsert and clean up on naming, please take a look when you have time

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/utils/config/TableConfigUtils.java:
```diff
@@ -116,8 +117,16 @@ public static TableConfig fromZNRecord(ZNRecord znRecord)
+    if (upsertConfigString != null) {
+      upsertConfig = JsonUtils.stringToObject(upsertConfigString, UpsertConfig.class);
+    }
+
```
(nit) extra empty line

Jackie-Jiang commented on pinot-common/src/test/java/org/apache/pinot/common/utils/config/TableConfigSerDeTest.java:
```diff
@@ -231,6 +234,17 @@ public void testSerDe()
     }
+    {
+      // with upsert config
+      UpsertConfig upsertConfig = new UpsertConfig(ImmutableList.of("pk"), "offset",
```
Prefer using `Collections.singletonList()`

Jackie-Jiang commented on pinot-common/src/test/java/org/apache/pinot/common/utils/config/TableConfigSerDeTest.java:
```diff
@@ -367,4 +381,12 @@ private void checkFieldConfig(TableConfig tableConfig) {
     assertNull(secondFieldConfig.getProperties());
   }
+
+  private void checkTableConfigWithUpsertConfig(TableConfig tableConfig, TableConfig tableConfigToCompare) {
```
Can you make this similar to other config check? Pass in a  table config and check the actual value inside the config. Then serialize and deserialize (both string and ZNRecord) and check again

Jackie-Jiang commented on pinot-spi/src/test/java/org/apache/pinot/spi/config/UpsertConfigTest.java:
```diff
@@ -0,0 +1,72 @@
+      fail();
+    } catch (RuntimeException ex) {}
+  }
+}
```
(nit) empty line

jamesyfshao commented:
hi @mcvsubbu we have updated the pr following the comments. Appreciate if you can take a look to see if anything else you feel we need to update further

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java:
```diff
@@ -0,0 +1,82 @@
+    _validUntilColumn = validUntilColumn;
+
+    Preconditions.checkState(_primaryKeyColumns.size() == 1,
+        "upsert feature support only one primary key column");
```
```suggestion
        "Upsert feature supports only one primary key column");
```

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java:
```diff
@@ -0,0 +1,82 @@
+    Preconditions.checkState(_primaryKeyColumns.size() == 1,
+        "upsert feature support only one primary key column");
+    Preconditions.checkState(StringUtils.isNotEmpty(_offsetColumn),
+        "upsert feature requires \"offsetColumn\" to be set ");
```
```suggestion
        "Upsert feature requires \"offsetColumn\" to be set ");
```

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java:
```diff
@@ -0,0 +1,82 @@
+    Preconditions.checkState(StringUtils.isNotEmpty(_offsetColumn),
+        "upsert feature requires \"offsetColumn\" to be set ");
+    Preconditions.checkState(StringUtils.isNotEmpty(_validFromColumn),
+        "upsert feature requires \"validFromColumn\" to be set");
```
```suggestion
        "Upsert feature requires \"validFromColumn\" to be set");
```

mcvsubbu commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java:
```diff
@@ -0,0 +1,82 @@
+    Preconditions.checkState(StringUtils.isNotEmpty(_validFromColumn),
+        "upsert feature requires \"validFromColumn\" to be set");
+    Preconditions.checkState(StringUtils.isNotEmpty(_validUntilColumn),
+        "upsert feature requires \"validUntilColumn\" to be set");
```
```suggestion
        "Upsert feature requires \"validUntilColumn\" to be set");
```

jamesyfshao commented:
> minor comments, other than that, lgtm thanks

merged the comments diff, thanks for checking on those details. Do you think any further changes you want? It would be great if you can change your review status from requesting-changes so I can start merging it in

mcvsubbu commented:
@jamesyfshao I am good with this PR, thanks.
